apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/asyncprofiler/JfrParser.java

```diff
@@ -0,0 +1,518 @@
+     *                       If {@code false}, will also resolve {@code Native}, {@code Kernel} and {@code C++} frames.
+     * @param stackFrames    The mutable list where the stack frames are written to.
+     *                       Don't forget to {@link List#clear()} the list before calling this method if the list is reused.
+     */
+    public void getStackTrace(long stackTraceId, boolean onlyJavaFrames, List<StackFrame> stackFrames) {
+        if (buffer == null) {
+            throw new IllegalStateException("getStackTrace was called before parse");
+        }
+        MappedByteBuffer buffer = this.buffer;
+        int position = buffer.position();
+        setPosition(buffer, stackTraceIdToFilePositions.get((int) stackTraceId));
+        long stackTraceIdFromFile = buffer.getLong();
+        assert stackTraceId == stackTraceIdFromFile;
+        buffer.get(); // truncated
+        int numFrames = buffer.getInt();
+        for (int i = 0; i < numFrames; i++) {
+            long method = buffer.getLong();
+            buffer.getInt(); // bci (always set to 0 by async-profiler)
+            byte frameType = buffer.get();
+            if (!onlyJavaFrames || isJavaFrameType(frameType)) {
+                LazyStackFrame lazyStackFrame = framesByFrameId.get(method);
+                if (lazyStackFrame.isIncluded(this)) {
+                    StackFrame stackFrame = lazyStackFrame.resolve(this);
+                    stackFrames.add(stackFrame);
+                }
+            }
+        }
+        setPosition(buffer, position);
+    }
+
+    private boolean isJavaFrameType(byte frameType) {
+        return isJavaFrameType[frameType];
+    }
+
+    private StringBuilder resolveSymbol(int pos, boolean replaceSlashWithDot) {
+        int currentPos = buffer.position();
+        setPosition(buffer, pos);
+        try {
+            return readUtf8String(replaceSlashWithDot);
+        } finally {
+            setPosition(buffer, currentPos);
+        }
+    }
+
+    private boolean isClassIncluded(CharSequence className) {
+        return WildcardMatcher.isAnyMatch(includedClasses, className) && WildcardMatcher.isNoneMatch(excludedClasses, className);
+    }
+
+    private static void setPosition(MappedByteBuffer buffer, int pos) {
+        // weird hack because of binary incompatibility introduced in Java 9
+        // the return type was changed from Buffer to MappedByteBuffer
+        // linking a method also incorporates the exact return type
+        ((Buffer) buffer).position(pos);
+    }
+
+    private StackFrame resolveStackFrame(int classId, int methodName) {
+        Symbol classNameSymbol = symbols.get(classIdToClassNameSymbolId.get(classId));
+        if (classNameSymbol.isClassNameIncluded(this)) {
+            String className = classNameSymbol.resolveClassName(this);
+            String method = symbols.get(methodName).resolve(this);
+            return new StackFrame(className, Objects.requireNonNull(method));
```

Comment:
> and use them with StringBuilders instead of Strings for class and method names, into which you can directly resolve from the file.

That's probably going to use more memory as the majority of symbols is likely to be unresolved. But pooling the strings and the frames separately could be a nice improvement.

Reply:
Or pool those StringBuilders, as suggested in https://github.com/elastic/apm-agent-java/pull/981#discussion_r365740422 😜